### ISSUE TITLE
hl_ - DOS attack by delegating tokens with `MAX_DELEGATES = 1024`

### ISSUE THREAD

**Issue by sherlock-admin4** [sherlock-audit/2024-06-velocimeter-judging#369]
hl_

High

# DOS attack by delegating tokens with `MAX_DELEGATES = 1024`

## Summary
DOS attack by delegating tokens with `MAX_DELEGATES = 1024`.

## Vulnerability Detail
Any user can delegate the balance of the tokens to anyone by calling `delegate`. The `VotingEscrow` contract implements a safety check of `MAX_DELEGATES = 1024`. This limit is intended to prevent excessively large arrays but does not fully address the gas cost issue for operations involving maximum delegation.

Given the current implementation, any user with 1024 delegated tokens takes approximately 23M gas to transfer/withdraw a token. However, this could exceed the limit of the EVM-compatible network (per mentioned in contest details) that the protocol intends to deploy on e.g. Optimism with a gas limit of 15M. An attacker could create a new address, delegate tokens to the victim's address, and accumulate the maximum allowed 1024 delegated tokens. By doing so, the attacker can create a situation where the victim faces extremely high gas costs for operations involving these tokens. In a case where the gas limit is hit by the attacker, the victim cannot withdraw/transfer/delegate.

## Impact
If an operation requires iterating over the entire list of delegated tokens, the gas cost can become prohibitively high, potentially causing the victim's transactions to fail or be blocked by exceeding the gas limit.

## Code Snippet
https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/VotingEscrow.sol#L1266

## Tool used
Manual Review

## Recommendation
Provide users with the option to opt-in or opt-out. Users will only accept delegated tokens if they choose to opt-in; otherwise, they can opt-out to decline any undesired delegated tokens.

Duplicate of #26
